test/python/transpiler/test_remove_final_measurements.py

```diff
@@ -0,0 +1,316 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Test RemoveFinalMeasurements pass"""
+
+import unittest
+
+from qiskit import ClassicalRegister, QuantumRegister, QuantumCircuit
+from qiskit.circuit.classicalregister import Clbit
+from qiskit.transpiler.passes import RemoveFinalMeasurements
+from qiskit.converters import circuit_to_dag
+from qiskit.test import QiskitTestCase
+
+
+class TestRemoveFinalMeasurements(QiskitTestCase):
+    """Test removing final measurements."""
+
+    def test_multi_bit_register_removed_with_clbits(self):
+        """Remove register when all clbits removed."""
+
+        def expected_dag():
+            q0 = QuantumRegister(2, "q0")
+            qc = QuantumCircuit(q0)
+            return circuit_to_dag(qc)
+
+        q0 = QuantumRegister(2, "q0")
+        c0 = ClassicalRegister(2, "c0")
+        qc = QuantumCircuit(q0, c0)
+
+        # measure into all clbits of c0
+        qc.measure(0, 0)
+        qc.measure(1, 1)
+
+        dag = circuit_to_dag(qc)
+        dag = RemoveFinalMeasurements().run(dag)
+
+        self.assertFalse(dag.cregs)
+        self.assertFalse(dag.clbits)
+        self.assertEqual(dag, expected_dag())
+
+    def test_register_kept_if_measured_clbit_busy(self):
+        """
+        A register is kept if the measure destination bit is still
+        busy after measure removal.
+        """
+
+        def expected_dag():
+            q0 = QuantumRegister(1, "q0")
+            c0 = ClassicalRegister(1, "c0")
+            qc = QuantumCircuit(q0, c0)
+            qc.x(0).c_if(c0[0], 0)
+            return circuit_to_dag(qc)
+
+        q0 = QuantumRegister(1, "q0")
+        c0 = ClassicalRegister(1, "c0")
+        qc = QuantumCircuit(q0, c0)
+
+        # make c0 busy
+        qc.x(0).c_if(c0[0], 0)
+
+        # measure into c0
+        qc.measure(0, c0[0])
+
+        dag = circuit_to_dag(qc)
+        dag = RemoveFinalMeasurements().run(dag)
+
+        self.assertListEqual(list(dag.cregs.values()), [c0])
+        self.assertListEqual(dag.clbits, list(c0))
+        self.assertEqual(dag, expected_dag())
```

Comment:
I like the explicit checks of the registers and bits before the full equality comparison of the DAGs - makes it easier to work out what's gone wrong in a test failure.